### DISCUSSION TITLE
Assertions for actions with modals?

### DISCUSSION THREAD

**Discussion by titantwentyone** [filamentphp/filament#8048] [Help]
Testing query again! Just working on a project which has a custom action which, in turn, generates a modal. I couldn't see anything in the docs (v2 or v3) nor browsing through the code to assert the modal content view and any data pased to it. Would a PR be welcome? My code from a table action

```
Action::make('event')
    ->modalContent(function($record) {
        return view('myview', [
            'data' => 'some data'
        ]);
    });
```
Just thinking something like:
```
livewire(ListModels::class)
    ->assertHasTableAction('event')
    ->assertTableActionHasModalContent('event', view('myview', [
            'data' => 'some data'
        ]));
```
And obviously replicated for Pages also.

Related but are you still adding functionality (i.e like this) to v2 while Livewire v3 is still in beta? I'm holding off on the upgrade to v3 for both for now but still actively developing with v2.

**Comment by danharrin** [comment]
Is there a difference with assetSee()?

> Related but are you still adding functionality (i.e like this) to v2 while Livewire v3 is still in beta?

Livewire 3 is now not in beta. I will merge bug fixes into v2 but probably not features. You can use a macro if you want to achieve this in v2.

**Reply by titantwentyone** [reply]
Was looking for something more in parity with `assertViewHas`, `assertViewIs` etc.

**Reply by titantwentyone** [reply]
Got some custom assertions done for my project. I did spot something though and I'll need to check in v3 but the following:

```
Action::make('event')
    ->modalContent(function($record) {
        return view('myview', [
            'data' => $record->value
        ]);
    });

livewire(SomeRelationManager::class, [
     'ownerRecord' => $myModel
])
->callTableAction('event', $myModel->items()->first());
```
Results in an error.
```
ErrorException: Attempt to read property "value" on null
```
I don't think the assertion takes into account that the table action could invoke `modalContent` with a closure. Not sure how much was transferred from v2 to v3 with regards to this and I'll take a look at some point. I've done a bit of digging but the error, I think, stems from TestsActions.php:L43 `shouldOpenModal` which uses `getModalContent` as part of a boolean check. I couldn't see any tests relating to this in either v2 or v3 for `modalContent`.
